File: backend/controller/dalamRangka_controller.go
package controller

import (
	"TaskPus/database"
	"TaskPus/helpers"
	"TaskPus/model"
	"TaskPus/structs"
	"net/http"

	"github.com/gin-gonic/gin"
)

func GetDalamRangka(c *gin.Context) {
	var dalamRangka []model.DalamRangka
	var response []structs.DalamRangkaResponse

	if err := database.DB.Find(&dalamRangka).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to fetch dalam rangka",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	for _, d := range dalamRangka {
		r := structs.DalamRangkaResponse{
			Id:   d.Id,
			Nama: d.Nama,
		}
		response = append(response, r)
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "List Data Dalam Rangka",
		Data:    response,
	})
}

func CreateDalamRangka(c *gin.Context) {
	var req = structs.CreateDalamRangkaRequest{}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, structs.ErrorResponse{
			Success: false,
			Message: "Validation Errors",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	dalamRangka := model.DalamRangka{
		Nama: req.Nama,
	}

	if err := database.DB.Create(&dalamRangka).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to create dalam rangka",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	r := structs.DalamRangkaResponse{
		Id:   dalamRangka.Id,
		Nama: dalamRangka.Nama,
	}

	c.JSON(http.StatusCreated, structs.SuccessResponse{
		Success: true,
		Message: "Dalam rangka created successfully",
		Data:    r,
	})
}

func GetDalamRangkaByID(c *gin.Context) {
	id := c.Param("id")
	var dalamRangka model.DalamRangka
	if err := database.DB.First(&dalamRangka, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "Dalam rangka not found",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}
	r := structs.DalamRangkaResponse{
		Id:   dalamRangka.Id,
		Nama: dalamRangka.Nama,
	}
	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Dalam rangka found",
		Data:    r,
	})
}

func UpdateDalamRangka(c *gin.Context) {
	var req = structs.UpdateDalamRangkaRequest{}
	id := c.Param("id")

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, structs.ErrorResponse{
			Success: false,
			Message: "Validation Errors",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}
	var dalamRangka model.DalamRangka
	if err := database.DB.First(&dalamRangka, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "Dalam rangka not found",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	dalamRangka.Nama = req.Nama

	if err := database.DB.Save(&dalamRangka).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to update dalam rangka",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	r := structs.DalamRangkaResponse{
		Id:   dalamRangka.Id,
		Nama: dalamRangka.Nama,
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Dalam rangka updated successfully",
		Data:    r,
	})
}

func DeleteDalamRangka(c *gin.Context) {
	id := c.Param("id")
	var dalamRangka model.DalamRangka
	if err := database.DB.First(&dalamRangka, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "Dalam rangka not found",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}
	if err := database.DB.Delete(&dalamRangka).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to delete dalam rangka",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}
	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Dalam rangka deleted successfully",
	})
}

File: backend/app/Http/Controllers/DalamRangkaController.php
<?php

namespace App\Http\Controllers;

use App\Models\DalamRangka;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Hash;
use Illuminate\Validation\ValidationException;
use Illuminate\Support\Facades\DB;
use Tymon\JWTAuth\Facades\JWTAuth;
use function App\Helpers\successResponse;
use function App\Helpers\errorResponse;

class DalamRangkaController extends Controller
{
    //
    public function GetDalamRangka()
    {
        $dalamRangka = DalamRangka::get();

        $response = $dalamRangka->map(function ($u) {
            return [
                'id'         => $u->id,
                'nama'       => $u->nama,
            ];
        });

        return response()->json([
            'success' => true,
            'message' => 'List Data DalamRangka',
            'data'    => $response
        ]);
    }

    public function CreateDalamRangka(Request $request)
    {
        $data = $request->validate([
            'nama'     => 'required|string',
        ]);

        $dalamRangka = DalamRangka::create($data);

        return successResponse('DalamRangka created successfully', [
            'id'         => $dalamRangka->id,
            'nama'       => $dalamRangka->nama,
        ]);
    }

    public function GetDalamRangkaById($id)
    {
        $dalamRangka = DalamRangka::findOrFail($id);

        return response()->json([
            'success' => true,
            'message' => 'DalamRangka found',
            'data'    => $this->DalamRangkaResponse($dalamRangka)
        ]);
    }

    public function UpdateDalamRangka(Request $request, $id)
    {
        $dalamRangka = DalamRangka::findOrFail($id);

        $data = $request->validate([
            'name'     => 'sometimes|required|string' . $dalamRangka->id,
        ]);

        $dalamRangka->update($data);

        return successResponse('DalamRangka updated successfully', $this->DalamRangkaResponse($dalamRangka));
    }

    public function DeleteDalamRangka($id)
    {
        $dalamRangka = DalamRangka::findOrFail($id);
        $dalamRangka->delete();

        return successResponse('DalamRangka deleted successfully');
    }

    private function DalamRangkaResponse(DalamRangka $dalamRangka)
    {
        return [
            'id'         => $dalamRangka->id,
            'nama'       => $dalamRangka->nama,
            'created_at' => $dalamRangka->created_at->format('Y-m-d H:i:s'),
            'updated_at' => $dalamRangka->updated_at->format('Y-m-d H:i:s'),
        ];
    }
}
